addons/project/models/project_task_stage_personal.py

```diff
@@ -12,7 +12,7 @@ class ProjectTaskStagePersonal(models.Model):
 
     task_id = fields.Many2one('project.task', required=True, ondelete='cascade', index=True, export_string_translation=False)
     user_id = fields.Many2one('res.users', required=True, ondelete='cascade', index=True, export_string_translation=False)
-    stage_id = fields.Many2one('project.task.type', domain="[('user_id', '=', user_id)]", ondelete='set null', export_string_translation=False)
+    stage_id = fields.Many2one('project.task.type', domain="[('user_id', '=', user_id)]", ondelete='restrict', export_string_translation=False)
```

Comment:
Note: `personal_stage_type_ids` override with `ondelete='restrict'`, avoid the confusion.